File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cs
using System;
using System.Collections.Generic;

public class Solution {
    private int[] dx = {1, -1, 0, 0};
    private int[] dy = {0, 0, 1, -1};
    private bool[,] visited;
    public int solution(string[] board) {
        int n = board.Length;
        int m = board[0].Length;
        
        int nx = 0; int ny = 0;
        
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                char c = board[i][j];
                if(c == 'R')  { nx = i; ny = j; break; }
            }
        }
        
        visited = new bool[n, m];
        int answer = BFS(board, nx, ny, 'G');
        return answer;
    }
    private int BFS(string[] board, int startX, int startY, char target)
    {
        int n = board.Length;
        int m = board[0].Length;
        
        Queue<(int, int, int)> q = new Queue<(int, int, int)>();
        q.Enqueue((startX, startY, 0));
        visited[startX, startY] = true;
        
        while(q.Count > 0)
        {
            var (x, y, d) = q.Dequeue();
            for(int i = 0; i < 4; i++)
            {
                int cx = x;
                int cy = y;
                while(true)
                {
                    int nx = cx + dx[i];
                    int ny = cy + dy[i];
                    
                    if(nx < 0 || nx >= n || ny < 0 || ny >= m) break;
                    if(board[nx][ny] == 'D') break;
                    
                    cx = nx;
                    cy = ny;
                }
                if(visited[cx, cy]) continue;
                if(board[cx][cy] == target) return d + 1;
                
                visited[cx, cy] = true;
                q.Enqueue((cx, cy, d + 1));
            }
        }
        return -1;
    }
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int dy[] = {-1, 0, 1, 0};
int dx[] = {0, 1, 0, -1};
int visited[101][101];

int sy, sx, x, y;

int solution(vector<string> board) {
    
    for(int i = 0; i < board.size(); i++)
    {
        for(int j = 0; j < board[0].size(); j++)
        {
            if(board[i][j] == 'R')
            {
                sy = i;
                sx = j;
            }
        }
    }
    
    queue<pair<int, int>> q;
    int cnt = 0;
    
    q.push({sy, sx});
    visited[sy][sx] = 1;
    
    while(q.size())
    {
        tie(y, x) = q.front();
        q.pop();
        
        if(board[y][x] == 'G') return visited[y][x] - 1;
        
        for(int i = 0; i < 4; i++)
        {
            int ny = y;
            int nx = x;
            
            while(true)
            {
                int ty = ny + dy[i];
                int tx = nx + dx[i];
                
                if(ty < 0 || tx < 0 || ty >= board.size() || tx >= board[0].size() )
                    break;
                if(board[ty][tx] == 'D') break;
                
                ny = ty;
                nx = tx;
            }
            if(!visited[ny][nx])
            {
                visited[ny][nx] = visited[y][x] + 1;
                q.push({ny, nx});
            }
        }
    }   
    
    return -1;
}